pkg/ingestor/parser/cyclonedx/parser_cyclonedx.go

```diff
@@ -427,33 +440,187 @@ func (c *cyclonedxParser) getAffectedPackages(ctx context.Context, vulnInput *mo
+			if err != nil {
+				return nil, fmt.Errorf("error searching version ranges for %s: %v", affect.Range, err)
+			}
+
+			for _, purl := range purls {
+				var addVulnDataErr error
+				viList, addVulnDataErr = c.AddVulnerabilityIngestionData(purl, vexData, vulnInput, viList)
+				if addVulnDataErr != nil {
+					return nil, fmt.Errorf("could not add Vulnerability information for %s: %v", pkgIdentifier, err)
+				}
+			}
+
 			continue
 		}
 		if affect.Version == "" {
 			return nil, fmt.Errorf("no version found for package ref %q", pkgRef)
 		}
-		vi := &assembler.VexIngest{
-			VexData:       &vexData,
-			Vulnerability: vulnInput,
-		}
 
 		// create guac specific identifier string using affected package name and version.
-		pkgID := guacCDXPkgPurl(pkdIdentifier, affect.Version, "", false)
-		pkg, err := asmhelpers.PurlToPkg(pkgID)
+		purl := guacCDXPkgPurl(pkgIdentifier, affect.Version, "", false)
+		var err error
+		viList, err = c.AddVulnerabilityIngestionData(purl, vexData, vulnInput, viList)
 		if err != nil {
-			return nil, fmt.Errorf("unable to create package input spec from guac pkg purl: %v", err)
+			return nil, fmt.Errorf("could not add Vulnerability information for %s: %v", pkgIdentifier, err)
 		}
-		vi.Pkg = pkg
-		viList = append(viList, *vi)
-		c.identifierStrings.PurlStrings = append(c.identifierStrings.PurlStrings, pkgID)
 	}
 
 	return &viList, nil
 }
 
+func (c *cyclonedxParser) AddVulnerabilityIngestionData(purl string, vexData model.VexStatementInputSpec, vulnInput *model.VulnerabilityInputSpec, viList []assembler.VexIngest) ([]assembler.VexIngest, error) {
+	pkg, err := asmhelpers.PurlToPkg(purl)
+	if err != nil {
+		return nil, fmt.Errorf("unable to create package input spec from guac pkg purl: %v", err)
+	}
+	vi := &assembler.VexIngest{
+		VexData:       &vexData,
+		Vulnerability: vulnInput,
+		Pkg:           pkg,
+	}
+	viList = append(viList, *vi)
+	c.identifierStrings.PurlStrings = append(c.identifierStrings.PurlStrings, purl)
+	return viList, nil
+}
+
+func (c *cyclonedxParser) findCDXPkgVersionIDs(ctx context.Context, pkgIdentifier string, versionRange string) ([]string, error) {
+	var matchingVersionPurls []string
+
+	// search for all packages that have a matching type, namespace, and name.
+	typeGUAC, namespace := "guac", "pkg"
+	pkgResponse, err := getPackages(ctx, c.gqlClient, model.PkgSpec{
+		Type:      &typeGUAC,
+		Namespace: &namespace,
+		Name:      &pkgIdentifier,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("error querying for packages1: %w", err)
+	}
+
+	pkgVersionsMap := map[string]string{}
+	var pkgVersions []string
+	for _, depPkgVersion := range pkgResponse.Packages[0].Namespaces[0].Names[0].Versions {
+		pkgVersions = append(pkgVersions, depPkgVersion.Version)
+		pkgVersionsMap[depPkgVersion.Version] = depPkgVersion.Id
+		pkgVersionsMap[depPkgVersion.Version] = guacCDXPkgPurl(typeGUAC, depPkgVersion.Version, "", false)
```

Comment:
`depPkgVersion` should contain a `purl` that can just be called and used instead of recreating using `guacCDXPkgPurl`

Reply:
I don't think that `depPkgVersion` contains a purl. The definition: https://github.com/guacsec/guac/blob/d95860c03e8956a89573ca47156737dddb4701ab/pkg/assembler/clients/generated/operations.go#L5248-L5253

Reply:
ah so it can. See my PR here: https://github.com/guacsec/guac/pull/1872/files#diff-22ea6f9d7837c7075a97304dc077bd4f4f345fa28f1f99acde3600d475d83ee0

If you want to add that and update your PR, I can rebase on it.

Reply:
You will have to run a make generate once you make that change and it will be usable.

Reply:
I can include the purl into the struct, but since the purl's actual value isn't being initialized in this PR (it is being initialized in #1872), the feature won't work correctly (and the tests will fail). Instead, I added a todo to do a small follow-up after both PRs are merged in.

Reply:
Oh, I see what you mean. The code is now updated with `Purl` from `depPkgVersion`.